Translate py

alfredats opened the pull request:
I made added a "PY.hs", extracted the FEEL functions into "FEELhelpers.hs" file to be used by both "JS.hs" & "PY.hs", and made some changes to the "app/Main.hs" and "app/Options.hs" files to enable python output. PY.hs is largely similar to JS.hs, except that functionality relating to typescript output is removed.

Currently it produces the appropriate python functions for the various tables within the README.md in the root folder. 

Would appreciate any feedback, as well as further direction.

alfredats committed:
Add functionality for Python output

- Added "src/PY.hs" for output of Python functions for dmnmd
- Extracted FEEL functions from "src/JS.hs" into a seperate file
  "FEELhelper.js"
- Added Python as one of the possible output formats in "app/Main.hs" and
  "app/Options.hs"

alfredats committed:
Oops. Forgot to add the changes to Main.hs and Options.hs

Changes described in previous commit

alfredats committed:
Fixed logical operators for output in Python

mengwong commented on languages/haskell/src/DMN/Translate/PY.hs:
```diff
@@ -0,0 +1,146 @@
+import DMN.Translate.FEELhelpers
+
+data PYOpts = PYOpts { propstyle :: Bool
+                     , typescript :: Bool }
```
I don't think python files are going to need typescript functionality. Is there some sort of gradual typing used in Python that is the approximate equivalent of what the JS module attempts to do for TS?

alfredats replied:
I left this in for a future step where we can enable strict typing for function arguments as described here: https://docs.python.org/3/library/typing.html

Perhaps I could remove it first and re-introduce it later on?

mengwong replied:
Yes, better to remove it for now.

mengwong commented on languages/haskell/app/Options.hs:
```diff
@@ -64,6 +64,7 @@ parseFileFormat :: OA.ReadM FileFormat
+    "py" -> return Py
     "md" -> return Md
     "xml" -> return Xml
     _    -> Left "Accepted file types are 'ts', 'js', 'xml', and 'md'."
```
This should also say 'py'

mengwong committed:
quasiquotation blocks close with "|]" -- table parser now ignores

mengwong committed:
add examples for higher order functions in output columns

this is NOT in the DMN spec!

mengwong commented on languages/haskell/src/DMN/Translate/FEELhelpers.hs:
```diff
@@ -0,0 +1,43 @@
+showFeel (FNullary (VS str))  = show str
+showFeel (FNullary (VN num))  = show num
+showFeel (FNullary (VB bool)) = toLower <$> show bool
+showFeel (FSection Feq  (VB rhs)) = "(x)=>x === "++ (toLower <$> show rhs)
```
the `FSection` cases here are intended to produce syntax for lambda functions.

In JS and TS, lambda functions are defined compactly as `(x)=>……`.

The Python version is a bit more verbose, a bit like `lambda x: ……`

This section of the code hasn't really been exercised until now – I have added an output column to the `language/haskell/README.md` that illustrates the return of lambda functions.

mengwong commented on languages/haskell/src/DMN/Translate/PY.hs:
suggest refactoring so that the lambda expressions output correctly for both the JS module and the PY module. You could add an argument to `showFeel` which produces the correct syntax for a lambda depending on JS vs PY; and the caller (JS vs PY) could provide that syntax when it calls `showFeel`.

alfredats committed:
Add functionality for lambda syntax output in Python

Changes made to Types.hs
 - Added 2 new types, FNLog and FNComp. These will be used in
   FEELhelpers.hs to generate the appropriate logcial and comparison
   operators for each output format.

Changes made to FEELhelpers.hs
 - Created lambdaHeader function to generate appropriate lambda syntax
   for JS ("(x)=>") and Python ("lambda x: ") respectively
 -  Created functions showFNLog and showFNComp that return the
   logical and comparison operators for JS/TS and Python respectively.
   These functions utilize the newly created datatypes FNLog and FNComp
   in Types.hs

alfredats committed:
Update PY.hs and JS.hs to use FEELhelpers.hs

Updated the language-translation files to utilize the showFNLog and
showFNComp functions from FEELhelpers.hs

alfredats committed:
Update showToJSON in app/Main.hs

I'm not sure if showToJSON is utilized in the XML generation. If it
does, I probably bricked something.

alfredats committed:
Update error msg for parseFileFormat in app/Options.hs

alfredats committed:
Update error message for outputTo function in Main.hs

mengwong commented on languages/haskell/src/DMN/Translate/PY.hs:
there are a number of functions which seem to be the same across JS.hs and PY.hs, like wrapParen, nonBlankCols, input_headers, comment_headers, etc. would it make more sense to put them into FEELhelpers so they don't get repeated?

mengwong commented on languages/haskell/src/DMN/Translate/PY.hs:
```diff
@@ -0,0 +1,142 @@
+
+toPY :: PYOpts -> DecisionTable -> String
+-- https://github.com/faylang/fay/wiki
+toPY jsopts dt =
```
perhaps the jsopts could be renamed to pyopts, for consistency.

mengwong commented:
there are a few things which i think could be improved, like there is still quite a lot of repeat code, like in the `mkIf` function, but this is working.

Also, now that we have a working version, it would be nice to have some tests to make sure the JS and PY output doesn't get broken by future changes.

I will make some of these changes directly, and I would like you to review my changes as they happen over the course of the next few commits...
